node/src/node/slave/spawn_handle.cpp

```diff
@@ -0,0 +1,81 @@
+#include "cocaine/detail/service/node/slave.hpp"
+
+#include "cocaine/detail/service/node/slave/machine.hpp"
+#include "cocaine/detail/service/node/slave/spawn_handle.hpp"
+#include "cocaine/detail/service/node/slave/state/handshaking.hpp"
+#include "cocaine/detail/service/node/slave/state/spawn.hpp"
+
+#include "cocaine/service/node/slave/error.hpp"
+
+#include <cocaine/logging.hpp>
+#include <cocaine/hpack/header.hpp>
+
+#include <blackhole/logger.hpp>
+
+namespace cocaine {
+namespace detail {
+namespace service {
+namespace node {
+namespace slave {
+
+spawn_handle_t::spawn_handle_t(std::weak_ptr<machine_t> _slave, std::weak_ptr<state::spawn_t> _spawning) :
+    slave(std::move(_slave)),
+    spawning(std::move(_spawning))
+{}
+
+void
+spawn_handle_t::on_terminate(const std::error_code&, const std::string& msg) {
+    //TODO: IMPLEMENT
+    assert(false);
+    throw 42;
+}
+
+void
+spawn_handle_t::on_ready() {
+    auto _spawning = spawning.lock();
+    auto _slave = slave.lock();
+    if(!_spawning || !_slave) {
+        throw cocaine::error_t("possible bug - spawning state or slave is dead");
+    }
+
+    //TODO: Use
+    _spawning->cancel();
+    std::error_code ec;
+    const size_t cancelled = _spawning->timer.cancel(ec);
+    if (ec || cancelled == 0) {
+        // If we are here, then the spawn timer has been triggered and the slave has been
+        // shutdowned with a spawn timeout error.
+        COCAINE_LOG_WARNING(_slave->log, "slave has been spawned, but the timeout has already expired");
+        return;
+    }
+
+    const auto now = std::chrono::high_resolution_clock::now();
+    COCAINE_LOG_DEBUG(_slave->log, "slave has been spawned in {} ms",
+                      std::chrono::duration<float, std::chrono::milliseconds::period>(now - start).count());
+
+    try {
+        // May throw system error when failed to assign native descriptor to the fetcher.
+        auto handshaking = std::make_shared<state::handshaking_t>(_slave, std::move(_spawning->handle));
+        _slave->migrate(handshaking);
+
+        handshaking->start(_slave->profile.timeout.handshake);
+    } catch (const std::exception& err) {
+        COCAINE_LOG_ERROR(_slave->log, "unable to activate slave: {}", err.what());
+
+        _slave->shutdown(error::unknown_activate_error);
+    }
+}
+
+void
+spawn_handle_t::on_data(std::string data) {
+    auto _slave = slave.lock();
+    if(_slave) {
+        _slave->output(std::move(data));
```

Comment:
The logic should be inversed: we move data not just because we don't need it but because the external API requires it to own, not borrow.